### ISSUE TITLE
Callable gets called even when "required" fails

### ISSUE THREAD

**Issue by akinuri** [bcit-ci/CodeIgniter#5964]
Examine the following code. If the field is empty, validation (the `required` check) fails and (I'm assuming) `min_length[5]` never gets checked. I get the "field is required" message.

```php
$this->form_validation->set_rules('phone', 'Phone', "required|min_length[5]");
```

Now, check this. The callable is called even when the field is empty (`required` fails). I get "too short" message instead of "field is required".

```php
$this->form_validation->set_rules(
    'phone', 'Phone',
    array(
        'required',
        array('min_length', function ($str) {
            return strlen($str) >= 5;
        })
    ),
    array("min_length" => "too short")
);
```
Why is that? Do callables get special treatment? I'd expect them to work the same. If the field is marked required, but the field is empty, why call the callable?

---

Also, it seems the callable is called even if the `required` is not set and the field is empty.

If `required` is not set, that means the field is optional. And if the field is empty, there's nothing to validate. So, again, why is the callable called? A validation function expects a value. If an optional field is empty, there shouldn't be a validation. Yet, the callable is called. This will just return false (since the field is empty) and mislead the user into thinking there's something wrong with the field/form.

In short, if a callable is set as a rule, it gets called anyway. This doesn't make much sense to me. I think, in the following cases

1. If `required` is **set**, and if the `required` check **fails**
2. If `required` is **not set**, and the field is **empty**

callable should not be called.

**Comment by akinuri** [comment]
One way to get around this is to do a manual check if the field is not empty and then set the rule, but that just feels wrong. A validator should support this natively.

**Comment by akinuri** [comment]
I've been examining the `Form_validation.php` file and found this in `_prepare_rules()`:

> * "Callbacks" are given the highest priority (always called),
> * followed by 'required' (called if callbacks didn't fail),
> * and then every next rule depends on the previous one passing.

So, indeed, callbacks/callables gets a special treatment. This seems like an issue to me.

What is the reason for this? If the `required` is not set, callback/callable should only be called if the field is not empty. However that's not the case. A custom validation function will try to validate an empty field (even though it's optional), and fail.

And if the `required` is set, but the field is empty, the callback/callable will still fire as if there's a value to validate. Isn't this a problem?

**Comment by akinuri** [comment]
I've tried to come up with a simple solution, but it's too messy. For the time being, I'll do manual checks (see above) to get around it, and probably write a custom validator (that mimics HTML elements' attributes, ValidityState, etc.) later.

**Comment by otengkwame** [comment]
@akinuri Reading through I think the last comment should be the default way to go. You set custom messages for specific validations. You don't expect it to work with the known validation rules that have been specified in the documentation. Understanding the documentation is also part of the implementation.

**Comment by narfbg** [comment]
Duplicate of #5077.
